backstory/src/test/java/com/google/sps/story/StoryEndingToolsTest.java

```diff
@@ -0,0 +1,209 @@
+   */
+  @Test
+  public void emptyInputForRemoveFragment() {
+    Assert.assertEquals(0, StoryEndingTools.removeSentenceFragmentAtEnd(EMPTY_TEXT).length());
+  }  
+
+  /**
+   * Check that output from removeSentenceFragmentAtEnd
+   * with sentence fragments as input is the right length.
+   */
+  @Test
+  public void fragmentInputForRemoveFragment() {
+    for (String inputWithFragment: FRAGMENTS) {
+      int lastSentenceEnder = -1;
+
+      for (String ender: SENTENCE_ENDERS) {
+        int lastIndex = inputWithFragment.lastIndexOf(ender);
+
+        if (lastIndex > lastSentenceEnder) {
+          lastSentenceEnder = lastIndex;
+        }
+      }
+
+      Assert.assertEquals(lastSentenceEnder + 1, StoryEndingTools.removeSentenceFragmentAtEnd(inputWithFragment).length());
+    }
+  }  
+
+  /**
+   * Check that output from removeSentenceFragmentAtEnd
+   * with complete sentences as input is the right length.
+   */
+  @Test
+  public void sentenceInputWithRemoveFragment() {
+    for (String sentence: COMPLETE_SENTENCES) {
+      Assert.assertEquals(sentence.length(), StoryEndingTools.removeSentenceFragmentAtEnd(sentence).length());
+    }
+  }
+
+  /**
+   * Check that output from endStory always ends with correct
+   * punctuation.
+   */
+  @Test
+  public void endStoryAlwaysEndsProperly() {
+    for (String input: INPUTS) {
+      int lastSentenceEnder = -1;
+      String inputEnded = StoryEndingTools.endStory(input);
+
+      for (String ender: SENTENCE_ENDERS) {
+        int lastIndex = inputEnded.lastIndexOf(ender);
+
+        if (lastIndex > lastSentenceEnder) {
+          lastSentenceEnder = lastIndex;
+        }
+      }
+
+      Assert.assertEquals(inputEnded.length() - 1, lastSentenceEnder);
+    }
+  }
+
+  /**
+   * Check that output from endStory removes sentence fragments if
+   * there are any.
+   */
+  public void ensureFragmentRemoved() {    
+    // if it's a fragment, then with the ending removed, it should
+    // be shorter than it was before
+    for (String inputWithFragment: FRAGMENTS) {
+      String output = StoryEndingTools.endStory(inputWithFragment);
+      // remove last sentence by removing punctuation then calling remove fragment
+      String outputWithoutEnding = StoryEndingTools.removeSentenceFragmentAtEnd(output.substring(0, output.length() - 1));
+
+      Assert.assertTrue(inputWithFragment.length() > outputWithoutEnding.length());    
+    }
+
+    // if it's a sentence, then with the ending removed, it should
+    // be same length as it was before
+    for (String sentence: COMPLETE_SENTENCES) {
+      String output = StoryEndingTools.endStory(sentence);
+      // remove last sentence by removing punctuation then calling remove fragment
+      String outputWithoutEnding = StoryEndingTools.removeSentenceFragmentAtEnd(output.substring(0, output.length() - 1));
+
+      Assert.assertEquals(sentence.length(), outputWithoutEnding.length());
+    }
+  }
+}
```

Comment:
Is it possible to reduce the amount of logic in these tests?

See: https://drive.google.com/file/d/1QbOCmyI7hwhCUlHJ6NeH0Jfk7oqxZoPB/view

Reply:
It is possible to reduce the amount of logic. I see what they mean about it being more about input-output pairs. As such, I think that I'm going to rewrite emptyInputForRemoveFragment, fragmentInputForRemoveFragment, and sentenceInputWithRemoveFragment to be more like this as well as ensureFragmentRemoved. For the others, I believe the logic is necessary because it's more difficult to test those (as they get and attach a random ending), and I think the logic is fairly simple and allows testing on more inputs.

Reply:
Actually, in an **update**, I changed all of the methods, and there should be substantially less logic, and hopefully all of it is simpler.

Reply:
Done